File: app/filetype.py
from pathlib import Path
from typing import Optional

from app.logging_setup import get_logger
from app.schemas import METADATA_EXTENSIONS

log = get_logger(__name__)

DICOM_PREAMBLE_BYTES = 128
DICOM_MAGIC = b"DICM"

PDF_MAGIC = b"%PDF-"

OLE_MAGIC = b"\xd0\xcf\x11\xe0\xa1\xb1\x1a\xe1"

ZIP_MAGIC = b"PK\x03\x04"

SNIFF_BYTES = 4096

KNOWN_EXTENSIONS = frozenset(METADATA_EXTENSIONS)


def name_extension(name: str) -> str:
    suffix = Path(name).suffix
    return suffix[1:].lower() if suffix else ""


def sniff_extension(head: bytes) -> Optional[str]:
    if not head:
        return None

    if (
        len(head) >= DICOM_PREAMBLE_BYTES + len(DICOM_MAGIC)
        and head[DICOM_PREAMBLE_BYTES : DICOM_PREAMBLE_BYTES + len(DICOM_MAGIC)]
        == DICOM_MAGIC
    ):
        return "dcm"

    if head.startswith(DICOM_MAGIC):
        return "dcm"

    if head.startswith(PDF_MAGIC):
        return "pdf"

    if head.startswith(OLE_MAGIC):
        return "doc"

    if head.startswith(ZIP_MAGIC) and b"word/" in head:
        return "docx"

    return None


def resolve_extension(name: str, head: bytes) -> str:
    named = name_extension(name)
    if named in KNOWN_EXTENSIONS:
        return named

    sniffed = sniff_extension(head)
    if sniffed is None:
        return named

    log.info(
        "file_type_sniffed", name=name, named_extension=named or "(none)",
        detected=sniffed,
    )
    return sniffed


def head_of(path) -> bytes:
    try:
        with open(path, "rb") as handle:
            return handle.read(SNIFF_BYTES)
    except OSError as exc:
        log.warning("file_head_unreadable", path=str(path), error=str(exc))
        return b""

File: frontend/src/lib/fileType.ts
import { canDeidentify } from '@/schemas/applicationFile'

export const SUPPORTED_FORMATS_LABEL = 'DICOM, PDF, or Word'

const DICOM_PREAMBLE_BYTES = 128
const DICOM_MAGIC = 'DICM'
const PDF_MAGIC = '%PDF-'
const OLE_MAGIC = [0xd0, 0xcf, 0x11, 0xe0, 0xa1, 0xb1, 0x1a, 0xe1]
const ZIP_MAGIC = [0x50, 0x4b, 0x03, 0x04]
const SNIFF_BYTES = 4096

function nameExtension(name: string): string {
  const dot = name.lastIndexOf('.')
  return dot === -1 ? '' : name.slice(dot + 1).toLowerCase()
}

function startsWithAscii(bytes: Uint8Array, ascii: string, offset = 0): boolean {
  if (bytes.length < offset + ascii.length) return false
  for (let i = 0; i < ascii.length; i += 1) {
    if (bytes[offset + i] !== ascii.charCodeAt(i)) return false
  }
  return true
}

function startsWithBytes(bytes: Uint8Array, magic: number[]): boolean {
  if (bytes.length < magic.length) return false
  for (let i = 0; i < magic.length; i += 1) {
    if (bytes[i] !== magic[i]) return false
  }
  return true
}

function sniffExtension(head: Uint8Array): string | null {
  if (head.length === 0) return null

  if (
    startsWithAscii(head, DICOM_MAGIC, DICOM_PREAMBLE_BYTES) ||
    startsWithAscii(head, DICOM_MAGIC)
  ) {
    return 'dcm'
  }

  if (startsWithAscii(head, PDF_MAGIC)) return 'pdf'

  if (startsWithBytes(head, OLE_MAGIC)) return 'doc'

  if (startsWithBytes(head, ZIP_MAGIC)) {
    const text = new TextDecoder('latin1').decode(head)
    if (text.includes('word/')) return 'docx'
  }

  return null
}

async function readHead(file: File, bytes = SNIFF_BYTES): Promise<Uint8Array> {
  const buffer = await file.slice(0, bytes).arrayBuffer()
  return new Uint8Array(buffer)
}

export async function isSupportedUpload(file: File): Promise<boolean> {
  if (canDeidentify(nameExtension(file.name))) return true
  return sniffExtension(await readHead(file)) !== null
}

export async function partitionBySupport(
  files: File[]
): Promise<{ supported: File[]; unsupported: File[] }> {
  const flags = await Promise.all(files.map(isSupportedUpload))
  const supported: File[] = []
  const unsupported: File[] = []
  files.forEach((file, index) => {
    ;(flags[index] ? supported : unsupported).push(file)
  })
  return { supported, unsupported }
}
